
#'Plot the abundance of Predicted G-quadruplexes
#'Relative to a Central Postiion
#'
#'Using a vector returned by getQuadCoveragePercentage(),
#'plot the percentage of peaks that have G-quadruplexes across
#'positions in sequences around the ChIP-SEQ peaks
#'
#'@param featurePercentages a numeric vector containing percentages such as
#' a vector returned by getQuadCoveragePercentage()
#'@param title a string specifying the title of the plot
#'
#'@return returns TRUE if no errors
#'
#'@example
#' reports <- findQuads(bedPath = "BRD4_reduced.bed", seqWidth = 75, assemblyVersion = "hg19")
#' qMatrix <- getQuadMatrix(quadReports = reports, seqWidth = 75)
#' quadCoveragePercentage <- getQuadCoveragePercentage(quadMatrix = qMatrix)
#' plotQuadPosition(quadCoveragePosition, "BRD4 G-Quad Abundance")
#'
#'@export
plotQuadPosition <- function (featurePercentages, title) {
  #calculate the bounds for the x axis
  leftBound <- as.integer(-1*floor(length(featurePercentages)/2))
  rightBound <- as.integer(ceiling(length(featurePercentages)/2) - 1)
  #display the plot
  plot(c(leftBound, rightBound),
       c(0, 100),
       type = "n",
       xlab = "Position",
       ylab = "Percentage")
  lines(x = leftBound:rightBound,
        type = "l",
        y = featurePercentages,
        col = "blue")
  abline(v = 0)
  title(title)
  return(TRUE)
}

#'Generate a Pie Chart of The percentage of ChIP peak nucleotides that
#'are methylated
#'
#'Using the dataframe of methylation data for ChIP peak nucleotides
#' generated by getMethylOverlap, create a pie chart showing the
#' percentage of peak sites that are methylated
#'
#'@param methylOverlapData a data frame with a coverage column specifying
#' the percentage of reads methylated. This column must be named coverage
#'
#'@example
#' overlap <- getMethylOverlap("MAZ_very_small_test.bed", "HEK293MethylData.bed")
#' plotMethylPercentage(overlap)
#'
#'@export
plotMethylPercentage <- function(methylOverlapData) {
  #calculate the average percentages of reads methylated at each position
  avgPercentMethyl <- mean(methylOverlapData$coverage)
  #define the labels and sections for the pie chart
  pieLabels <-c(paste(round(avgPercentMethyl, 4), "%"),
                paste(round(100 - avgPercentMethyl, 4), "%"))
  sections <- c(avgPercentMethyl, 100 - avgPercentMethyl)
  #display the plot
  pie(sections, pieLabels,
      col = c("green", "blue"),
      cex = 0.7)
  title("Avg Percent Reads Methylated", cex = 0.5)
  legend("topright", c("Methylated", "Not Methylated"),
         cex = 0.53, fill= c("green", "blue"))
  return(TRUE)
}

#'Plot the percentage of sites methylated across multiple datasets in
#'a bar graph
#'
#'From a list of vectors containing the percentages of reads methylated
#'at different sites in ChIP peaks from the same ChIP peak dataset,
#'but potentially different methylation datasets
#'
#'@param methylCoverageList a list of vectors of methylation percentages
#'
#'@return returns TRUE if there are no errors
#'
#'@export
plotMultipleMethylPercentage <- function(methylCoverageList, lineNames, title, xlabel) {
  #calculate the average percentages of reads methylated at each position
  #for each dataset
  avgPercentages <-lapply(methylCoverageList, mean)
  avgPercentages <- unlist(avgPercentages)
  names(avgPercentages) <- lineNames
  #display the plot
  barplot(avgPercentages,
          ylim = c(0, max(avgPercentages) + 1),
          xlab = xlabel,
          ylab = "Average Coverage Percentage",
          main = title)
  return(TRUE)
}
